janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/CustomAnalyzer.java

```diff
@@ -0,0 +1,121 @@
+import org.apache.lucene.analysis.standard.StandardAnalyzer;
+import org.janusgraph.core.schema.Mapping;
+import org.janusgraph.core.schema.Parameter;
+import org.janusgraph.diskstorage.indexing.KeyInformation;
+import org.janusgraph.graphdb.types.ParameterType;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * @author Olivier binda (olivier.binda@wanadoo.fr)
+ */
+
+/**
+ * JanusGraph may open different index stores and needs an analyzer for each
+ * so an analyzer should be provided to build IndexWriters/IndexReaders that are used when an index is opened and documents are added
+ * BUT a janusGraph may add a new indexable field, with a custom tokenizing process through Parameters, at any moment, which means that it should be possible
+ * to map any field to the corresponding custom analyzer, though indexreaders are already in use.
+ * <p>
+ * A way to achieve that would be to use a delegating analyzer that delegates Tokenizing for a field to
+ * the corresponding Analyzer (whose class you get through a Mapping)
+ */
+public class CustomAnalyzer extends DelegatingAnalyzerWrapper {
+    private String store;
+    private KeyInformation.IndexRetriever informations;
+
+    private Map<String, Analyzer> analyzers = new HashMap<>();
+
+    public CustomAnalyzer(String store, KeyInformation.IndexRetriever informations, ReuseStrategy fallbackStrategy) {
+        super(fallbackStrategy);
+        this.store = store;
+        this.informations = informations;
+    }
+
+    @Override
+    protected Analyzer getWrappedAnalyzer(String fieldName) {
+        Mapping mapping = null;
+        String stringAnalyzerString = null;
+        String textAnalyzerString = null;
+
+
+        Parameter[] parameters = null;
+        Class<?> clazz = null;
+
+
+        KeyInformation keyInformation = informations.get(store, fieldName);
+        if (keyInformation != null) {
+            parameters = keyInformation.getParameters();
+            clazz = keyInformation.getDataType();
+        }
+
+        if (parameters != null) {
+            // if mapping isn't present in parameters, we use Mapping.DEFAULT
+            mapping = ParameterType.MAPPING.findParameter(parameters, Mapping.DEFAULT);
+            stringAnalyzerString = ParameterType.STRING_ANALYZER.findParameter(parameters, null);
+            textAnalyzerString = ParameterType.TEXT_ANALYZER.findParameter(parameters, null);
+        }
+      
+        String analyzerString = null;
+
+        if (clazz != null && String.class.isAssignableFrom(clazz) && mapping != null) {
+            // at the moment, we only try to support custom analyzers for string data.
+            // everything else falls through a StandardAnalyzer as was the case before
+            switch (mapping) {
+                case TEXTSTRING:
+                    throw new RuntimeException("TextString is an unsupported mapping for string data & custom analyzers");
```

Comment:
Instead of Runtime, can you use PermanentBackendException, because it's a problem of configuration

Reply:
I tried to but Idea is complaining because of the signature of getWrappedAnalyzer which is not supposed to throw checked exceptions and PermanentBackEndException extends Exception and not RuntimeException

I'm afraid the code won't compile if I do so (or is there another way ?)